### PR TITLE
Rashuai/boost cuda TopK performance

### PR DESCRIPTION
**Pull request by RandySheriffH**
Implement http://anilshanbhag.in/static/papers/gputopk_sigmod18.pdf to address issue https://github.com/microsoft/onnxruntime/issues/1830.
Some performance statistics from GTX 1050 Ti, Ubuntu 17.04, cuda 10 are:

| Platform | 16 * top 5 of 128 | 16 * top 50 of 1280 |  16 * top 500 of 12800  | 16 * top 5000 of 128000 |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| PyTorch  | 10 us | 11  | 40  | 3050  |
| ORT(now) | 4 us  | 4 | 12  | 540  |
| ORT(before)  | 90 us  | 120  | 500  | 540  |

### CODE REVIEW

File: onnxruntime/test/providers/cpu/math/topk_op_test.cc [hunk 1]
```diff
@@ -456,5 +456,35 @@ TEST(TopKOperator, SortedSelection) {
...
+  std::iota(expected_vals.begin(), expected_vals.end(), 9000.0f);
+  std::reverse(expected_vals.begin(), expected_vals.end());
+  std::vector<int64_t> expected_indices(1000, 0);
+  std::iota(expected_indices.begin(), expected_indices.end(), 9000);
+  std::reverse(expected_indices.begin(), expected_indices.end());
+  std::vector<int64_t> expected_dimensions = {1000};
+  RunTest(11, 1000, input_vals, input_dimensions, expected_vals, expected_indices, expected_dimensions, false, 0, 1, 1);
+}
+
+TEST(TopKOperator, BitArrayBigTopKSorted)
```
Comment (skottmckay) [inline onnxruntime/test/providers/cpu/math/topk_op_test.cc]: >BitArrayBigTopKSorted [](start = 19, length = 21)

BigArrayBigTopKSorted?

**Comment by skottmckay** [comment]
Can you please provide some performance numbers to quantify the improvement?

File: onnxruntime/core/providers/cuda/math/topk_impl.cu [hunk 2]
```diff
@@ -38,30 +334,47 @@ __global__ void ExcludeOutput(int64_t* output_i, int64_t K, int64_t dimension) {
...
+    auto input_value_buffer = kernel->GetScratchBuffer<int64_t>(dimension);
+    auto output_value_buffer = kernel->GetScratchBuffer<int64_t>(dimension);
+    auto input_key = input_key_buffer.get();
+    auto output_key = output_key_buffer.get();
+    auto input_value = input_value_buffer.get();
+    auto output_value = output_value_buffer.get();
+    size_t temp_bytes = 0;
+    CUDA_RETURN_IF_ERROR(cub::DeviceRadixSort::SortPairs(nullptr, temp_bytes, input_key, output_key, input_value, output_value, dimension));
+    auto temp_storage_buffer = kernel->GetScratchBuffer<char>(temp_bytes);
+    auto temp_storage = temp_storage_buffer.get();
```
Comment (skottmckay) [inline onnxruntime/core/providers/cuda/math/topk_impl.cu]: >auto [](start = 4, length = 4)

Please use const/*/& with 'auto' where applicable. As this is a pointer, 'auto *' provides more clarity.

File: onnxruntime/core/providers/cuda/math/topk_impl.cu [hunk 3]
```diff
@@ -38,30 +334,47 @@ __global__ void ExcludeOutput(int64_t* output_i, int64_t K, int64_t dimension) {
...
+    auto blocksPerGridD = (int)(ceil(static_cast<float>(dimension) / GridDim::maxThreadsPerBlock));
+    auto blocksPerGridK = (int)(ceil(static_cast<float>(K) / GridDim::maxThreadsPerBlock));
+    for (int64_t i = 0; i < N; i++) {
```
Comment (skottmckay) [inline onnxruntime/core/providers/cuda/math/topk_impl.cu]: nit: Casing doesn't match naming convention: blocks_per_grid_D

File: onnxruntime/test/providers/cpu/math/topk_op_test.cc (hunk 1 above)
Comment (RandySheriffH) [reply onnxruntime/test/providers/cpu/math/topk_op_test.cc]: Sorry just a typo, renaming to BigArray...

**Comment by RandySheriffH** [comment]
> Can you please provide some performance numbers to quantify the improvement?
Listed in PR summary.

**Pull request merged**